modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java

```diff
@@ -535,4 +604,179 @@ private static void validateParsedStatement(
+                        .whenComplete((res, ex) -> {
+                            if (ex != null) {
+                                txWrapper.rollback();
+                                cursorFuture.completeExceptionally(ex);
+                                cancelAll(ex);
+
+                                return;
+                            }
+
+                            if (parsedResult.queryType() == SqlQueryType.DML || parsedResult.queryType() == SqlQueryType.DDL) {
+                                txWrapper.commitImplicit();
+                            }
+
+                            cursorFuture.complete(res);
+                        });
+            } catch (Exception e) {
+                cursorFuture.completeExceptionally(e);
+
+                cancelAll(e);
+            }
+        }
+
+        private CompletableFuture<AsyncSqlCursor<List<Object>>> executeParsedStatementAndWaitPrefetch(
+                QueryTransactionWrapper txWrapper,
+                ParsedResult parsedResult,
+                Object[] dynamicParams
+        ) {
+            CompletableFuture<Void> prefetchFut = new CompletableFuture<>();
+            QueryCancel cancel = new QueryCancel();
+
+            return executeParsedStatement(schemaName, parsedResult, txWrapper, cancel, dynamicParams, new PrefetchCallback(prefetchFut))
+                    // Wait for the prefetch, otherwise the user will be able to execute
+                    // a dependent statement before the current one completes.
+                    .thenCompose(cur -> prefetchFut.thenApply(ignored -> cur));
+        }
+
+        private List<ScriptStatementParameters> splitDynamicParameters(List<ParsedResult> parsedResults, Object[] params) {
+            int totalDynamicParams = parsedResults.stream().mapToInt(ParsedResult::dynamicParamsCount).sum();
+
+            List<ScriptStatementParameters> results = new ArrayList<>(parsedResults.size());
+
+            validateDynamicParameters(totalDynamicParams, params);
+
+            int pos = 0;
+
+            for (ParsedResult result : parsedResults) {
+                Object[] params0 = Arrays.copyOfRange(params, pos, pos + result.dynamicParamsCount());
+
+                results.add(new ScriptStatementParameters(result, params0));
+
+                pos += result.dynamicParamsCount();
+            }
+
+            return Collections.unmodifiableList(results);
+        }
+
+        private void cancelAll(Throwable cause) {
+            for (ScriptStatementParameters parameters : statements) {
+                CompletableFuture<AsyncSqlCursor<List<Object>>> fut = parameters.cursorFuture;
+
+                if (fut == null || fut.isDone()) {
+                    continue;
+                }
+
+                fut.completeExceptionally(new SqlException(
+                        EXECUTION_CANCELLED_ERR,
+                        "The script statement execution was canceled due to an error in the previous statement.",
```

Comment:
nitpicking: I would omit first `statement` in the message

Reply:
Fixed, thanks.